keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -221,31 +231,53 @@ class SqlArtifactRepository(
     }
   }
 
-  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> =
-    if (isRegistered(artifact.name, artifact.type)) {
-      sqlRetry.withRetry(READ) {
-        jooq
-          .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
-          .from(ARTIFACT_VERSIONS)
-          .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
-          .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
-          .apply { if (artifact.statuses.isNotEmpty()) and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray())) }
-          .fetch()
-          .getValues(ARTIFACT_VERSIONS.VERSION)
-      }
-        .sortedWith(artifact.versioningStrategy.comparator)
-        .also { versions ->
-          // FIXME: remove special handling for Docker
-          // FIXME: limit the sql query, not the returned list (once we'll have a native sorting mechanism)
-          return if (artifact is DockerArtifact) {
-            filterDockerVersions(artifact, versions, limit)
-          } else {
-            versions.subList(0, Math.min(versions.size, limit))
+  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> {
+    // sanity checks
+    if (!isRegistered(artifact.name, artifact.type)) {
+      throw NoSuchArtifactException(artifact)
+    }
+
+    if (artifact.sortBy == BRANCH_AND_TIMESTAMP && artifact.branch == null) {
+      throw InvalidArtifactSpecException(
+        "The ${artifact.type} artifact ${artifact.name} specifies sorting by branch and timestamp, but no branch is defined in the config.")
+    }
+
+    val versions = sqlRetry.withRetry(READ) {
+      jooq
+        .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
+        .from(ARTIFACT_VERSIONS)
+        .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
+        .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
+        .apply {
+          if (artifact.statuses.isNotEmpty()) {
+            and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray()))
+          }
+
+          // When sorting "natively" by branch and timestamp, delegate sorting and limiting to the database
+          if (artifact.sortBy == BRANCH_AND_TIMESTAMP) {
+            // TODO: should we also be comparing the repo with what's configured for the app in front50?
+            and(ARTIFACT_VERSIONS_BRANCH.eq(artifact.branch))
+              .and(ARTIFACT_VERSIONS.CREATED_AT.isNotNull)
```

Comment:
so here, we are excluding versions which don't have a created at time yet, right? I'd love if there was some way to know that we are excluding some here, but it doesn't seem possible to add a log message telling us that. I'd like us to easily know that a version exists, but doesn't have enough metadata to be used in case a user complains of slowness

Reply:
I mean... We could add a separate query in this function to check for that and log, if we think it's worth the extra cost. Thoughts?

Reply:
good call Emily.. another potential way to test it out would be to count the # of versions which returning from this call and compare it to the limit to see the differences if any.

Reply:
@emjburns I'm planning to follow up with a PR to address https://github.com/spinnaker/keel/issues/1565 so we get metrics when we're missing artifact metadata.